nnf/tseitin.py

```diff
@@ -0,0 +1,63 @@
+"""Transformations using the well-known Tseitin encoding
+<https://en.wikipedia.org/wiki/Tseytin_transformation>
+
+The Tseitin transformation converts any arbitrary circuit to one in CNF in
+polynomial time/space. It does so at the cost of introducing new variables
+(one for each logical connective in the formula).
+"""
+
+from nnf import NNF, Var, And, Or, memoize, Internal
+
+
+aux_count = 0
+
+
+def Aux() -> Var:
+    global aux_count
+    aux_count += 1
+    return Var("aux_%d" % aux_count)
+
+
+def to_CNF(theory: NNF, skip_simplification: bool = False) -> And[Or[Var]]:
+    """Convert an NNF into CNF using the Tseitin Encoding.
+
+    :param theory: Theory to convert.
+    :param skip_simplification: If true, the final CNF will not be simplified
+    """
+
+    clauses = []
+
+    @memoize
+    def process_node(node: NNF) -> Var:
+
+        if isinstance(node, Var):
+            return node
+
+        assert isinstance(node, Internal)
+
+        aux = Aux()
+        children = {process_node(c) for c in node.children}
+
+        if isinstance(node, And):
+            clauses.append(Or({~c for c in children} | {aux}))
+            for c in children:
+                clauses.append(Or({~aux, c}))
+
+        elif isinstance(node, Or):
+            clauses.append(Or(children | {~aux}))
+            for c in children:
+                clauses.append(Or({~c, aux}))
+
+        else:
+            raise TypeError(node)
+
+        return aux
+
+    root = process_node(theory)
+    clauses.append(Or({root}))
+    theory = And(clauses)
+
+    if skip_simplification:
+        return theory
+    else:
+        return theory.simplify()
```

Comment:
This won't work until #8 is implemented, and even then I wouldn't use it because it can change the structure of the sentence (e.g. `And({Or({Var(1)})}).simplify() == Var(1)`). It also doesn't remove as many clauses as it could.

I think it's best to add something like this right after the definition of `children` (not tested):
```python
if any(~var in children for var in children):
    if isinstance(node, And):
        clauses.append(Or({~aux}))
    else:
        clauses.append(Or({aux}))
    return aux
```

(I'd also remove the `skip_simplification` argument unless there's a compelling need for it)

Reply:
This friggen thing's been a royal pain all morning. So many intertwined issues (from `is_CNF` being checked in random places) and different requirements on that call.

What you propose works, at least for that spot. But now we face failures of a different nature -- when the variable gets wiped, it doesn't appear in the theory (`vars()` is computed via walking through, rather than a static list). This has broader implications on counting and simplification (if variables disappear, then the counts may not be accurate).

So in [`test_tseitin`](https://github.com/QuMuLab/python-nnf/blob/tseitin/test_nnf.py#L760), it both fails to have the right number of models, and the `satisfied_by` method fails as well (since the model it's checking doesn't have a value for the variable it removed).

Reply:
Relenting a bit on the test...
https://github.com/QuMuLab/python-nnf/commit/7c56919eaaef6c5d8e37b5ff4cf4dbbde77b83f3

...and will tackle the variable mismatch down the road (#11 )

Reply:
Ouch!

The principled solution would be the new equivalence test, but that isn't ready yet.

For now, you could use a function that supplements models given a series of names. There's [a specialized internal version of that](https://github.com/QuMuLab/python-nnf/blob/e296aa46ccafd393dd35d00a648b25493d3e769f/nnf/__init__.py#L1157) somewhere already, but it's useful in more places. Maybe something like this as a top-level function in `nnf/__init__.py`:

```python
def complete_models(models: t.Iterable[Model], names: t.Frozenset[Name]) -> t.Iterator[Model]:
    diff = None
    for model in models:
        if diff is None:
            diff = names - model.keys()
        for supplement in all_models(diff):
            new = model.copy()
            new.update(supplement)
            yield new
```

And then you could test like this:

```python
models = list(complete_models(T.models(), sentence.vars() | T.vars()))
for mt in models:
    assert sentence.satisfied_by(mt)
assert len(models) == sentence.model_count()
```

Reply:
I posted it after you resolved the conversation, but do you think this fix is worth the trouble?

Reply:
Hopefully easy enough. Ran out of runway for today (will be back tonight). Latest on this branch fails when there is a 0-var theory.

Reply:
But the suggested change is in / tested (to a limited extent).

Reply:
I think I've fixed it.